core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java

```diff
@@ -4177,6 +4178,109 @@ RelOptFixture checkDynamicFunctions(boolean treatDynamicCallsAsConstant) {
         .check();
   }
 
+  @Test void testJoinWithConstantRowOnRight() {
```

Comment:
How about a test where some of the `VALUE`s are NULL?

Reply:
These tests are designed to validate the generated plans for a specified query. I've incorporated output oriented tests into the join.iq file, and couple of them generate null values in the context of a left outer join. If you think any test cases are missing in join.iq, please let me know, and I'll be sure to add them. Thank you.

Reply:
equi-joining with a tuple that has a null value for one of the ON fields should produce an empty set. Is there a test about that?

Reply:
I have added two test cases with null values as input. Please take a look and let me know. Thank you.

Reply:
I meant a test like this one: `select e.empno, e.ename, c.* from emp e cross join (select 5, NULL) c`
This test seems to produce an incorrect plan because you just substitute the NULL for the corresponding column:

```
LogicalProject(EMPNO=[$0], ENAME=[$1], EXPR$0=[$9], EXPR$1=[$10])
  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8], $f9=[5], $f10=[null:NULL])
    LogicalTableScan(table=[[CATALOG, SALES, EMP]])
```

I believe that the result should be the empty set.

Reply:
@mihaibudiu  Thanks for the review, the above query doesn't result in an empty subset as it is a cross join. I have added the test case and made sure that the results are same with and without the optimization (by removing the rules  and executing the test case).